### PR TITLE
Add support partial wildcard to find marshaler

### PR DESCRIPTION
**Pull request by duck8823**
I tried to add CustomMarshaler for `multipart/form-data`. but http client ( such as curl ) usually add generated boundary strings to a header `Content-Type`  (ex. `Content-Type: multipart/form-data boundary=MarshalerForRequest`). 
Unfortunately, the current marshaler_registory only supports perfect matching strings or overall wildcard.
I propose this PR that add support partial wildcard to finding marshaler in marshaler_registory.

This makes it possible to...
```golang
mux := runtime.NewServeMux(
		runtime.WithMarshalerOption("multipart/form-data; boundary=*", &CustomMarshaler{}),
	)
```

It is not limited for `multipart/form-data`, can also be used for `audio/*`, `image/*` and etc.

Thank you.

### CODE REVIEW

**Comment by duck8823** [comment]
@achew22 
Great thanks for your review! 
Please let me confirm if I got it.

I think to create new function that returns a ServeMuxOption.
e.x `runtime.WithMarshalerRegexOption` 

1. Is my understanding correct ? 
2. Which one is better ?
  a. The new function's argument is Regex. 
     How to use this... 
     ```
     headerRegex := regexp.Compile("application\/.*")
     mux := runtime.NewServeMux(
       runtime.WithMarshalerRegexOption(headerRegex, &CustomMarshaler{})
     )
     ```
   b. or arguement is string with only partical wildcard.
      ```     
      mux := runtime.NewServeMux(
        runtime.WithMarshalerWildcardOption("application/*", &CustomMarshaler{})
      )
      ```
   
    I like b than a, because simply usage.

Thanks.

File: runtime/marshaler_registry_test.go [hunk 1]
```diff
@@ -105,3 +162,23 @@ type dummyEncoder struct{}
 func (dummyEncoder) Encode(interface{}) error {
 	return errors.New("not implemented")
 }
+
+type dummyMarshaler2 struct {
+	label int
```
[the remark this answers is not in the document]
Comment (duck8823) [reply runtime/marshaler_registry_test.go]: To distinguish. When there is no label and initialize `var marshalers [4]dummyMarshaler` ,  `marshalers[0] == marshalers[1]` returns `true`.

**Comment by duck8823** [comment]
@achew22 
Can you see this?

**Comment by yugui** [comment]
> I think to create new function that returns a ServeMuxOption.
> e.x `runtime.WithMarshalerRegexOption`
> 
> Is my understanding correct ?

I don't think so.  Maybe he is saying that we need to have another method than `marshalerRegistry.add`.

You can determine if the given mime type has a partial wildcard or not, then call the right method of `marshalerRegistry`.

**Comment by johanbrandhorst** [comment]
@duck8823 Are you still interested in finishing this?

**Comment by duck8823** [comment]
Sorry, but I'm busy with another work...
Can I close this PR myself?

**Comment by johanbrandhorst** [comment]
I'll close it for you. Hopefully someone can come along and implement this if they need it.

**Pull request closed without merging**